### DISCUSSION TITLE
TCP from Teach Pendant

### DISCUSSION THREAD

**Discussion by domrice** [HALRobotics/Beta#122] [Q&A] [answered]
Hello!

How do I set the the tool TCP in HAL to the values from the teach pendant?

As I imagine, it can be set with a Frame where I put the values into x, y, and z. But how do I get the frame from the robot flange? It is defined as part (even though it shows a frame) that is not convertable into a plane.
![TCPfromFrame](https://user-images.githubusercontent.com/58233897/156757331-9ce6d461-753b-4041-9ca9-d22e21b936a9.png)
 

Thanks in advance!

**Comment by sebandraos** [comment, marked as answer]
@DominikReisach I'm not sure I understand the context of the question completely but hopefully one of the answers below is useful:
1. To define the TCP, that would be the EndFrame of a Tool, defined relative to the Flange of the robot. That will then be transformed to the correct position when the Tool is Attached to the Robot. [This tutorial](https://docs.hal-robotics.com/Grasshopper/2-Cell/Contents.html#24-create-a-tool) might be useful to get a handle on this.
2. If you need to read data from the teach pendant and convert it to HAL formats then you can use the overloads of Frame to select the correct formalism for your robot manufacturer. [This tutorial](https://docs.hal-robotics.com/Grasshopper/2-Cell/Contents.html#25-calibrate-a-reference-or-tool) should help with that.
3. There is a subtle but important distinction between two Reference input types in the Framework. References in Frames are other Frames/Planes. This is to help define Frames relative to each other to construct data but they are collapsed into a single Frame and that hierarchy isn't retained. References in Parts, References, Targets etc. are a means of defining a structure that persists during simulation and execution e.g. when you need to re-calibrate a Reference or when a Part can move and your Targets need to move with it.

Hopefully one of those is the answer you are looking for but let us know if that's not clear.

**Reply by domrice** [reply]
@sebandraos Thanks for your reply. Most likely, your second point is the correct solution. I will try that tomorrow.

I am sorry for not beeing aware of that tutorial. Though, that line in "2.4 Create a Tool" lead me astray, because the links didn't work correctly in my browswer:
"In a [future tutorial] we will see how to import that frame from calibration data in your [Controller] but to keep things simple here I’m going to use the point at the tip of the pencil in my model."
